viabel/approximations.py

```diff
@@ -547,4 +547,142 @@ def _pth_moment(self, var_param, p):
+        ----------
+        dim : `int`
+            dimension of the underlying parameter space
+        k : 'int'
+            number of low rank
+        """
+        self._rs = npr.RandomState(seed)
+        self._pattern = _get_low_rank_mu_sigma_pattern(dim, k)
+        self._k = k
+        super().__init__(dim, self._pattern.flat_length(True), True, True)
+
+    def init_param(self):
+        init_param_dict = dict(mu=np.zeros(self.dim),
+                               log_sigma=np.ones(self.dim),
+                               low_rank=self._rs.randn(self.dim,self._k))
+        return self._pattern.flatten(init_param_dict)
+
+
+
+    def sample(self, var_param, n_samples, seed=None):
+        my_rs = self._rs if seed is None else npr.RandomState(seed)
+        param_dict = self._pattern.fold(var_param)
+        z = my_rs.randn(n_samples,self._k)
+        epsilon = my_rs.randn(n_samples, self.dim)
+        D_exp = np.exp(param_dict['log_sigma'])
+        B = param_dict['low_rank']
+
+        return param_dict['mu'] + np.dot(z,B.T) + D_exp * epsilon
+
+    def _entropy(self, var_param):
+        param_dict = self._pattern.fold(var_param)
+        B = param_dict['low_rank']
+        D_exp = np.exp(2 * param_dict['log_sigma'])
+        Sigma_det = matrix_determinant_lemma(D_exp, B)
+
+        return 0.5 * self.dim * (np.log(2 * np.pi ) + 1) + 0.5 * np.log(Sigma_det)
 
+    def _kl(self, var_param0, var_param1):
+        param_dict0 = self._pattern.fold(var_param0)
+        param_dict1 = self._pattern.fold(var_param1)
+        mean_diff = param_dict0['mu'] - param_dict1['mu']
+
+        B0 = param_dict0['low_rank']
+        D0 = np.exp(2 * param_dict0['log_sigma'])
+        Sigma0_det = matrix_determinant_lemma(D0, B0)
+
+        B1 = param_dict1['low_rank']
+        D1 = np.exp(2 * param_dict1['log_sigma'])
+        D1_inv = np.diag(1/D1)
+
+        Sigma1_det = matrix_determinant_lemma(D1, B1)
+
+        #By the Woodbury formula
+        D1_invB = np.dot(D1_inv, B1)
+        I_BDB_inv = np.linalg.inv(np.eye(self._k) + np.dot(B1.T, D1_invB))
+        Sigma_inv = D1_inv - D1_invB @ I_BDB_inv @ D1_invB.T
+
+
+        #KL divergence
+        Sigma_log_diff = np.log(Sigma1_det / Sigma0_det)
+        Mean_sigma = mean_diff.T @ Sigma_inv @ mean_diff
+        Sigma_trace = matrix_trace_lemma(D0,B0,D1,B1)
+        return .5 * (Sigma_log_diff - self.dim + Mean_sigma + Sigma_trace)
+
+
+    def log_density(self,var_param, x):
+        if x.ndim == 1:
+            x = x[np.newaxis, :]
+        param_dict = self._pattern.fold(var_param)
+        mean = param_dict['mu']
+        B = param_dict['low_rank']
+        D_exp = np.exp(2*param_dict['log_sigma'])
+        D_inv = np.diag(1/D1)
+
+        Sigma_det = matrix_determinant_lemma(D_exp, B)
+
+        # By the Woodbury formula
+        D_invB = D_inv @ B
+        I_BDB_inv = np.linalg.inv(np.eye(self._k) + np.dot(B.T, D_invB))
+        Sigma_inv = D_inv - D_invB @ I_BDB_inv @ D_invB.T
```

Comment:
If you are careful about multiplying diff by each term in `Sigma_inv` you can avoid any $O(k^2)$ operations / space requirements